### PR TITLE
[DO NOT MERGE] Prepare container rootfs with containerdPrepare method

### PR DESCRIPTION
**Pull request by adarsh-zededa**
Changes in this PR:
1) With the loaded image, `ctrCreate()` creates a new container.
2)  New Task is created using `ctrCreateNewTask()` at the time of container creation (`ctrCreate()`)
3) `containersRoot` and `containerRootfsPath` is updated to the path where the snapshot will be extracted by `ctrCreateNewTask()` step
4) `getContainerConfigs()` returns config details from container's spec
5) `ctrRm()` deletes task and container for the given id. This will eventually cleanup unused snapshots.

Pending changes:
1) Test cases
2) How to handle image clean-up?

### CODE REVIEW

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 1]
```diff
@@ -36,44 +39,45 @@ type KeyValue struct {
 
 // getContainerPath return the path to the root of the container. This is *not*
 // necessarily the rootfs, which may be a layer below
-func getContainerPath(id string) string {
-	return path.Join(containersRoot, id)
+func getContainerPath(containerID string) string {
+	return path.Join(containersRoot, CtrdServicesNamespace, containerID)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: Something isn't quite right here. The namespace isn't precisely under `/run/containerd/<namespace>` but is split:

* the namespace itself is under `/run/containerd/runc/<namespace>`, with the container as `/run/containerd/runc/<namespace>/<container_name>/`
* the rootfs is somewhere else, pointed to by `/run/containerd/runc/<namespace>/<container_name>/state.json`, particularly the property `config.rootfs`. For example, here it is on a default docker Ubuntu install:

```
root@ip-172-31-15-153:~# cat /run/containerd/runc/default/alpine2/state.json | jq -r '.config.rootfs'
/run/containerd/io.containerd.runtime.v1.linux/default/alpine2/rootfs
root@ip-172-31-15-153:~# ls /run/containerd/io.containerd.runtime.v1.linux/default/alpine2/rootfs
bin  dev  etc  home  lib  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
```

The location is configurable, as is the runtime config file.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 2]
```diff
@@ -36,44 +39,45 @@ type KeyValue struct {
...
 }
 
 // getContainerRootfs return the path to the root of the container filesystem
-func getContainerRootfs(id string) string {
-	return path.Join(getContainerPath(id), containerRootfsPath)
+func getContainerRootfs(containerID string) string {
+	return path.Join(getContainerPath(containerID), containerRootfsPath)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: As above, this might or might not be correct. It depends on what we are looking for. In general, though, the config will be above the rootfs.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 3]
```diff
@@ -36,44 +39,45 @@ type KeyValue struct {
...
 	}
 	// successful, so return the list of images we imported
-	names := make(map[string]string)
+	imageNames := make([]string, 0)
 	for _, tag := range imgs {
-		names[tag.Name] = string(tag.Target.Digest)
+		imageNames = append(imageNames, tag.Name)
 	}
-	return names, nil
+	return imageNames, nil
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: Why convert this from a `map[string]string` with the hashes to just a `[]string`? We lose some information that way. True, we don't use it yet, but why lose the info?

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 4]
```diff
@@ -197,3 +288,61 @@ func getRktPodManifest(PodManifestFile string) (RktPodManifest, error) {
 	}
 	return manifest, nil
 }
+
+func loadContainer(containerID string) (containerd.Container, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: I suggest you either return a pointer `*containerd.Container` so you can return `nil` if the container doesn't exist, or have a special `error` that you can use to indicate it. There is a difference, which we will use, if the container does not exist (e.g. `ctrRm` doesn't think that is an error, as it is the desired end-state) vs if we had an error.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 5]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+	log.Infof("ctrRm %s\n", containerID)
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return fmt.Errorf("Could not build containerd client. %v", err.Error())
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	container, err := loadContainer(containerID)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: As below, we need to distinguish between an error vs the container does not exist. If it is gone, then we are fine, we got where we wanted, so `return nil`.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 6]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+	if err != nil {
+		return err
+	}
+	task, err := container.Task(ctrdCtx, cio.NewAttach(cio.WithStdio))
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: This could cause issues, since it uses a fifo, and only one attach is allowed at a time. See [here](https://godoc.org/github.com/containerd/containerd/cio#Attach):

> There should only be one reader for a task's IO set because fifo's can only be read from one reader or the output
> will be sent only to the first reads

Can you pass it `nil`? We aren't looking at the output here, so who cares?

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 7]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+
+	if len(status.EnvVariables) > 0 {
+		envVars := buildEnvVarsSlice(status.EnvVariables)
+		ociOpts = append(ociOpts, oci.WithEnv(envVars))
+	}
+
+	if len(ociOpts) > 0 {
+		containerOpts = append(containerOpts, containerd.WithSpec(&ociSpec, ociOpts...))
+	}
+	image, err := getImage(imageName)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: This seems wasteful. You already have a client and context in hand, why not just use it? It is a single line:

```go
	image, err := ctrdClient.GetImage(ctrdCtx, imageName)
```

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 8]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+	if err != nil {
+		return "", fmt.Errorf("Unable to fetch image %v. %v", imageName, err.Error())
+	}
+
+	containerSnapshot := fmt.Sprintf("%s-snapshot", containerID)
+	containerOpts = append(containerOpts, containerd.WithImage(image))
+	containerOpts = append(containerOpts, containerd.WithNewSnapshot(containerSnapshot, image))
+	containerOpts = append(containerOpts, containerd.WithNewSpec(oci.WithImageConfig(image)))
+	_, err = ctrdClient.NewContainer(
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: This is a bit wasteful. We ignore the returned `Container` here, and then recreate it in `ctrCreateNewTask()` below, right down to a new client. Why not just keep the response and use it?

```go
cont, err := ctrdClient.NewContainer(
//...
)
//...
task, err := container.NewTask(ctrdCtx, cio.NewCreator(cio.WithStdio))
```

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 9]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+		return "", fmt.Errorf("Could not create a new task for container: %v. %v", containerID, err.Error())
+	}
+	return containerID, nil
+}
+
+// ctrCreateNewTask - Creates a new task for the container.
+// This step is responsible for setting up rootfs of the container under /run/containerd/<RunTime>/<NameSpace>/<Container id>/
+// NOTE: As per containerd's flow, to start a container, first a task is created which sets up container's rootfs,
+// then the task is started which invokes runc binary with the rootfs. Since we are not using runc, we need to stop at task create step.
+func ctrCreateNewTask(containerID string) (containerd.Task, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: As above, this can be removed.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 10]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+		return nil, err
+	}
+
+	return container.NewTask(ctrdCtx, cio.NewCreator(cio.WithStdio))
 }
 
 // ctrPrepare prepare an existing container
 func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
-	ctrID := uuid.NewV4().String()
+	containerID := buildContainerID(status.DisplayName)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: I like your abstracting this out into a func.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 11]
```diff
@@ -197,3 +288,61 @@ func getRktPodManifest(PodManifestFile string) (RktPodManifest, error) {
...
+	for envName, envValue := range envMap {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", envName, envValue))
+	}
+
+	return envVars
+}
+
+func buildContainerID(displayName string) string {
+	uuid := uuid.NewV4().String()
+	return fmt.Sprintf("%s-%s", displayName, uuid)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: This is good as a function. Do we need to include the displayName in the container name? I would suggest we include just the generated UUID, and if we need anything else, include it as a label on the container, see [this option](https://godoc.org/github.com/containerd/containerd#WithContainerLabels)

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 12]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
 
-	// TODO:
-	// 1. create a container using containerd - remember that containerd does not
-	//    create a name for it, so we need to do so. Docker uses a hash of the config.
-	//    We can just create a UUID.
-	// 2. create a task using containerd - this will create the snapshot
-	// 3. include any necessary mountpoints
-	if err = ctrCreate(ctrID, imageHash); err != nil {
-		return ctrID, fmt.Errorf("failed to create container %s: %v", ctrID, imageHash)
+	if containerID, err := ctrCreate(containerID, imageName, status); err != nil {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: This is strange and non-idiomatic. `containerD := ctrCreate(containerID,...)`? We should not be overriding the var, especially as it is the same value. Linter should catch this. Just make it `if err = ctrCreate(...); err != nil {`

File: pkg/pillar/scripts/device-steps.sh [hunk 13]
```diff
@@ -21,7 +21,7 @@ AGENTS0="logmanager ledmanager nim nodeagent"
 AGENTS1="zedmanager zedrouter domainmgr downloader verifier identitymgr zedagent baseosmgr wstunnelclient"
 AGENTS="$AGENTS0 $AGENTS1"
 TPM_DEVICE_PATH="/dev/tpmrm0"
-
+XDG_RUNTIME_DIR=/var/persist/tmp
```
Comment (deitch) [inline pkg/pillar/scripts/device-steps.sh]: Why is this in here? Let's not mix up PRs.

File: pkg/pillar/scripts/device-steps.sh [hunk 14]
```diff
@@ -64,6 +64,7 @@ if [ -d $TMPDIR ]; then
 fi
 mkdir -p $TMPDIR
 export TMPDIR
+export XDG_RUNTIME_DIR
```
Comment (deitch) [inline pkg/pillar/scripts/device-steps.sh]: Same as above. Let's not mix up PRs.

File: pkg/pillar/vendor/github.com/docker/distribution/registry/api/errcode/errors.go [hunk 15]
```diff
@@ -207,11 +207,11 @@ func (errs Errors) MarshalJSON() ([]byte, error) {
 	for _, daErr := range errs {
 		var err Error
 
-		switch daErr.(type) {
```
Comment (deitch) [inline pkg/pillar/vendor/github.com/docker/distribution/registry/api/errcode/errors.go]: Let's not mix up PRs. If this is not relevant to what we are doing, put it in a separate PR, please.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 16]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+		return fmt.Errorf("Unable to delete container: %v. %v", containerID, err.Error())
+	}
 	log.Infof("ctrRm done\n")
 	return nil
 }
 
 // ctrCreate create a new container but do not start it
-func ctrCreate(id, imageHash string) error {
-	return nil
+func ctrCreate(containerID, imageName string, status *types.DomainStatus) (string, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: Hmm. I see the only thing we are using `status` for in this func is to get the env vars, which is a `map[string]string`. Could we just pass the `status.EnvVariables` to this func as a `map[string]string`? That would make this func entirely eve-independent, and really easy to test.

Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: And yes, I am acutely aware that I passed it through in the previous PR. It is your good work here that made me see the possibilities for improving it. :-)

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 17]
```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+	container, err := loadContainer(containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	return container.NewTask(ctrdCtx, cio.NewCreator(cio.WithStdio))
 }
 
 // ctrPrepare prepare an existing container
 func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: Same here. The only thing we use `status` for is to get:

* `status.DisplayName string`
* `status.EnvVariables map[string]string`

If we pass those 2 explicitly to `ctrPrepare()`, then the func has zero eve dependencies, and becomes much easier to test.

Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Actually, take it one step further. Here is the beginning of `ctrPrepare()`:

```go
func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
	ociFilename, err := utils.VerifiedImageFileLocation(containerImageSha256)
 	log.Infof("ociFilename %s sha %s", ociFilename, containerImageSha256)
```

We only use `containerImageSha256` to calculate the `ociFilename`. Move that back out to the calling function in `domainmgr.go`, passing it just the filename, and we can get rid of importing `lf-edge/eve/pkg/pillar/utils`. Then there will be zero dependencies.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 18]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/oci"
+	"github.com/opencontainers/runtime-spec/specs-go"
```
[the remark this answers is not in the document]
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Done.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 1 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Yeah, you are right. I missed runtime in the path. Corrected that.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 2 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Corrected.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 3 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: NewContainer() requires image as containerd.Image{} type, but ImageImport() returned image as images.Image{} type. So to convert that, I did getImage() with the imageName which returned the image as containerd.Image{} type. 
But as you mentioned, this is clumsy as the same image is fetched twice. So on digging into container code for a simpler way to convert that.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 9 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Removed.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 12 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Corrected.

File: pkg/pillar/scripts/device-steps.sh (hunk 13 above)
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: There was an issue with mount path resolution. So to handle that these changes were added.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 16 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Done.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 17 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Done.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 6 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: As per discussion with @rvs, I've removed create task step, instead will me manually mounting container's snapshot.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 7 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Corrected.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 8 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Removed.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: It still is strange. Maybe `gofmt` isn't moving them? In any case, it is commonly accepted practice to put all standard library parts in the first section, then a blank line, then all of the external (`github.com/whatever...`) after. Can we move these down to the second `import` section ?

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 2 above)
Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Did you test it? The path to the container might not be the root of the path to the rootfs.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 6 above)
Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Got it. What about the `cio.NewAttach()`?

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 19]
```diff
@@ -90,70 +112,143 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
...
+		return fmt.Errorf("ctrCreate: Could not build rootfs of container: %v. %v", containerID, err.Error())
+	}
 	return nil
 }
 
 // ctrPrepare prepare an existing container
-func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
-	ctrID := uuid.NewV4().String()
-	ociFilename, err := utils.VerifiedImageFileLocation(containerImageSha256)
+func ctrPrepare(ociFilename string, envVars map[string]string, noOfDisks int) (string, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: Oh, this is _so_ nice... :-)

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 1 above)
Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: The rootfs still might be in a different location. Need to read the `state.json`, or know how we will configure the containers on creation.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 3 above)
Comment (deitch) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Got it. This one didn't matter all that much, but nice to see

File: pkg/pillar/scripts/device-steps.sh (hunk 13 above)
Comment (rvs) [reply pkg/pillar/scripts/device-steps.sh]: But I think those are fixed in a different way -- so lets remove these now @adarsh-zededa

File: pkg/pillar/vendor/github.com/docker/distribution/registry/api/errcode/errors.go (hunk 15 above)
Comment (rvs) [reply pkg/pillar/vendor/github.com/docker/distribution/registry/api/errcode/errors.go]: Yeah -- lets remove it @adarsh-zededa

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Rearranged.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 1 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Changed the rootfs location under /persist/runx/

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 4 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Done.

File: pkg/pillar/cmd/domainmgr/containerd.go (hunk 5 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/domainmgr/containerd.go]: Done.

File: pkg/pillar/scripts/device-steps.sh (hunk 13 above)
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: Removed.

File: pkg/pillar/vendor/github.com/docker/distribution/registry/api/errcode/errors.go (hunk 15 above)
Comment (adarsh-zededa) [reply pkg/pillar/vendor/github.com/docker/distribution/registry/api/errcode/errors.go]: These changes were generated as part of go.mod changes.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 20]
```diff
@@ -188,12 +272,210 @@ func getRktPodManifest(PodManifestFile string) (RktPodManifest, error) {
...
+	err = ioutil.WriteFile(filepath.Join(appDir, imageConfigFilename), []byte(imageConfigJSON), 0666)
+	if err != nil {
+		return fmt.Errorf("Exception while writing image info to %v/%v. %v", appDir, imageConfigFilename, err)
+	}
+
+	err = ioutil.WriteFile(filepath.Join(appDir, pidFilename), []byte(container.ID()), 0666)
+	if err != nil {
+		return fmt.Errorf("Exception while writing container pid to %v/%v. %v", appDir, pidFilename, err)
+	}
+	_, err = executeShellCommands(mountCommands)
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: Can we replace the exec-ing out to a shell with using direct library calls? [Mount](https://godoc.org/golang.org/x/sys/unix#Mount) should help

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 21]
```diff
@@ -188,12 +272,210 @@ func getRktPodManifest(PodManifestFile string) (RktPodManifest, error) {
...
+	for _, m := range mounts {
+		mountCommands = append(mountCommands, &command{
+			name: "mount",
+			args: []string{"-t", m.Type, m.Source, target, "-o", strings.Join(m.Options, ",")},
+		})
+	}
+	return mountCommands
+}
+
+func executeShellCommands(commands []*command) ([]string, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: As above, we should be able to get rid of this if we use direct commands like [Mount](https://godoc.org/golang.org/x/sys/unix#Mount)

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 22]
```diff
@@ -162,23 +252,17 @@ func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string
...
+	env := imageInfo.Config.Env
 
-	return
+	for k, v := range userEnvVars {
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
+	return mountpoints, execpath, workdir, env, nil
 }
 
 func getRktPodManifest(PodManifestFile string) (RktPodManifest, error) {
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: We should be able to delete this.

File: pkg/pillar/cmd/domainmgr/containerd.go [hunk 23]
```diff
@@ -188,12 +272,210 @@ func getRktPodManifest(PodManifestFile string) (RktPodManifest, error) {
...
+	}
+	return results, nil
+}
+
+func deleteBundle(containerID string) {
+	os.RemoveAll(getContainerPath(containerID))
+}
+
+func isContainerNotFound(e error) bool {
+	return strings.HasSuffix(e.Error(), ": not found")
```
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/containerd.go]: It is unfortunate that we have to do this, and that containerd doesn't have a special error type.
